testing/file_test/file_test_base.cpp

```diff
@@ -462,13 +462,62 @@ struct SplitState {
   int file_index = 0;
 };
 
+// Replaces the content keywords.
+//
+// TEST_NAME is the only content keyword at present, but we do validate that
+// other names
+static auto ReplaceContentKeywords(llvm::StringRef filename,
+                                   std::string* content) -> ErrorOr<Success> {
+  static constexpr llvm::StringLiteral Prefix = "[[@";
+
+  auto keyword_pos = content->find(Prefix);
+  // Return early if not finding anything.
+  if (keyword_pos == std::string::npos) {
+    return Success();
+  }
+
+  // Construct the test name by getting the base name without the extension,
+  // then removing any "fail_" or "todo_" prefixes.
+  llvm::StringRef test_name = filename;
+  if (auto last_slash = test_name.rfind("/");
+      last_slash != llvm::StringRef::npos) {
+    test_name = test_name.substr(last_slash + 1);
+  }
+  if (auto ext_dot = test_name.find("."); ext_dot != llvm::StringRef::npos) {
+    test_name = test_name.substr(0, ext_dot);
+  }
+  while (test_name.consume_front("fail_") || test_name.consume_front("todo_")) {
+  }
```

Comment:
Bit surprised we want to support repeated `fail_` prefixes? Maybe just:

```suggestion
  test_name.consume_front("todo_");
  test_name.consume_front("fail_");
```

Or even:
```suggestion
  test_name.consume_front("todo_");
  test_name.consume_front("fail_");
  test_name.consume_front("todo_");
```

Reply:
IIUC we only recognize `fail_` if it's at the start, so
```suggestion
  test_name.consume_front("fail_");
  test_name.consume_front("todo_");
```

Reply:
We have both fail_todo_ and todo_fail_. So I've dropped the loop on request, but we should still handle both variations.

Reply:
And to be clear, these mean different things:

TODO, currently passes, should fail
TODO, currently fails, should succeed

Reply:
New approach works for me.